src/events/messageCreate.ts

```diff
@@ -97,9 +38,224 @@ export function makeHandler(
+	return logger[level](
+		JSON.stringify({
+			handle: `${msg.author.username}#${msg.author.discriminator}`,
+			user: msg.author.id,
+			message: msg.id,
+			...payload
+		})
+	);
+}
+
+// The only allowed exceptions are final reply errors or initial database access failures
+export async function onDirectMessage(
+	msg: Message<PrivateChannel>,
+	database: Public<DatabaseWrapperPostgres>,
+	decks: DeckManager,
+	challonge: WebsiteInterface,
+	participantRole: ParticipantRoleProvider,
+	bot: Client
+): Promise<void> {
+	let tournaments = await database.getPendingTournaments(msg.author.id);
+	if (tournaments.length > 1) {
+		const out = tournaments.map(t => t.name).join(", ");
+		log("info", msg, { event: "pending multiple", tournaments: out });
+		await reply(
+			msg,
+			`You are registering in multiple tournaments. Please register in one at a time by unchecking the reaction on all others.\n${out}`
+		);
+		return;
+	}
+	if (tournaments.length === 1) {
+		const tournament = tournaments[0];
+		log("info", msg, { event: "confirm start", tournament: tournament.id });
+		try {
+			await verifyDeckAndConfirmPending(msg, tournament, database, decks, challonge, participantRole, bot);
+		} catch (error) {
+			log("info", msg, { event: "confirm fail", tournament: tournament.id, error: error.message });
+			await reply(
+				msg,
+				`Must provide a valid attached \`.ydk\` file OR valid \`ydke://\` URL for **${tournament.name}**!`
+			);
+			await reply(msg, error.message);
+		}
+		return;
+	}
+
+	tournaments = await database.getConfirmedTournaments(msg.author.id);
+	if (tournaments.length > 1) {
+		const out = tournaments.map(t => t.name).join(", ");
+		log("info", msg, { event: "confirmed multiple", tournaments: out });
+		await reply(
+			msg,
+			`You're trying to update your deck for a tournament, but you're in multiple! Please choose one by dropping and registering again.\n${out}`
+		);
+		return;
+	}
+	if (tournaments.length === 1) {
+		const tournament = tournaments[0];
+		log("info", msg, { event: "update start", tournament: tournament.id });
+		try {
+			await verifyDeckAndUpdateConfirmed(msg, tournament, database, decks, bot);
+		} catch (error) {
+			log("info", msg, { event: "update fail", tournament: tournament.id, error: error.message });
+			await reply(
+				msg,
+				`Must provide a valid attached \`.ydk\` file OR valid \`ydke://\` URL for **${tournament.name}**!`
+			);
+			await reply(msg, error.message);
+		}
+		return;
+	}
+
+	log("verbose", msg, { event: "no context", content: msg.content });
+	await reply(
+		msg,
+		"Emcee's documentation can be found at https://github.com/AlphaKretin/emcee-tournament-bot/wiki. If you're trying to sign up for a tournament, make sure you've registered and I'll let you know how to proceed."
+	);
+}
+
+// Throws on any problem with the deck, and the exception payload should be sent to the user
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
```

Comment:
It is, but I found that this was really extra if the type signatures ever change. This is an internal function shared between the two code paths so there really are no doubts about what the types are and they will become a problem in the downstream code that actually uses it instead of creating more busywork here.